keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,859 @@
+-->
+
+## Design Details
+
+<!-- This section should contain enough information that the specifics
+of your change are understandable. This may include API specs (though
+not always required) or even code snippets. If there's any ambiguity
+about HOW your proposal will be implemented, this is the place to
+discuss them. -->
+
+We will expose a endpoint `/discovery` that will support JSON, and
+protobuf and gzip compression. The endpoint will serve the aggregated
+discovery document for all types that a Kubernetes cluster supports.
+
+### API
+
+The contents of this endpoint will be an `APIGroupList`, which is the
+same type that is returned in the discovery document at `/apis`. The
+`APIGroupList` contains a list of `APIGroup` types which includes a
+list of
+[`GroupVersionForDiscovery`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1071)
+types. We will modify this `GroupVersionForDiscovery` type to include
+a list of
+[`APIResource`](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1080).
+This list of `APIResource` is what is currently published at the
+`/apis/<group>/version` endpoint and is what will be aggregated into
+the new endpoint that we publish.
+
+The endpoint will also publish an ETag calculated based on a hash of
+the data for clients.
+
+This is approximately what the new API will look like (conflicting names will be renamed)
+
+```go
+// APIGroupList is a list of APIGroup, to allow clients to discover the API at
+// /apis.
+type APIGroupList struct {
+	TypeMeta `json:",inline"`
+	// groups is a list of APIGroup.
+	Groups []APIGroup `json:"groups" protobuf:"bytes,1,rep,name=groups"`
+}
+
+// APIGroup contains the name, the supported versions, and the preferred version
+// of a group.
+type APIGroup struct {
+	TypeMeta `json:",inline"`
+	// name is the name of the group.
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// versions are the versions supported in this group.
+	Versions []GroupVersionForDiscovery `json:"versions" protobuf:"bytes,2,rep,name=versions"`
+	// preferredVersion is the version preferred by the API server, which
+	// probably is the storage version.
+	// +optional
+	PreferredVersion GroupVersionForDiscovery `json:"preferredVersion,omitempty" protobuf:"bytes,3,opt,name=preferredVersion"`
+
+ 	// ServerAddresssByClientCIDRs will be removed for the new Discovery API
+	//ServerAddressByClientCIDRs []ServerAddressByClientCIDR `json:"serverAddressByClientCIDRs,omitempty" protobuf:"bytes,4,rep,name=serverAddressByClientCIDRs"`
+}
+
+// GroupVersion contains the "group/version" and "version" string of a version.
+// It is made a struct to keep extensibility.
+type GroupVersionForDiscovery struct {
+	// groupVersion specifies the API group and version in the form "group/version"
+	GroupVersion string `json:"groupVersion" protobuf:"bytes,1,opt,name=groupVersion"`
```

Comment:
unnecessary?

Reply:
I think either a group or this groupVersion string would be useful here if we are just manipulating this one object without the parent APIGroup object.

Reply:
I'm not a big fan of redundancy though